src/app/search/search.component.ts

```diff
@@ -443,14 +396,254 @@ export class SearchComponent implements OnInit {
+  /**===============================================
+   *                Helper Functions
+   * ==============================================
+   */
+  mapFriendlyValueNames(key, subBucket) {
+    if (key === 'tags.verified' || key === 'private_access') {
+      return this.friendlyValueNames.get(key).get(subBucket.toString());
+    } else {
+      return subBucket;
+    }
+  }
+  filterEntry() {
+    this.workflowHits = [];
+    this.toolHits = [];
+    for (const hit of this.hits) {
+      /**TODO: this is not good, make it faster.../
+       */
+      hit['_source'] = this.providerService.setUpProvider(hit['_source']);
+      if (hit['_type'] === 'tool') {
+        this.toolHits.push(hit);
+      } else if (hit['_type'] === 'workflow') {
+        this.workflowHits.push(hit);
+      }
+    }
+  }
+  parseOrderBy(key): any {
+    let order: any;
+    if (this.sortModeMap.has(key)) {
+      switch (this.sortModeMap.get(key).SortBy) {
+        case true: {
+          order = {
+            _count: this.sortModeMap.get(key).CountOrderBy ? 'asc' : 'desc'
+          };
+          break;
+        }
+        case false: {
+          order = {
+            _term: this.sortModeMap.get(key).AlphabetOrderBy ? 'asc' : 'desc'
+          };
+          break;
+        }
+        default: {
+          order = {
+            _count: 'desc'
+          };
+          break;
+        }
+      }
+    } else {
+      order = {
+        _count: 'desc'
+      };
+    }
+    return order;
+  }
+  /**
+   * This handles selection of one filter, either taking it out from the list of active filters
+   * or adding it if not present
+   * @param category
+   * @param categoryValue
+   */
+  handleFilters(category: string, categoryValue: string) {
+    if (this.filters.has(category) && this.filters.get(category).has(categoryValue)) {
+      this.filters.get(category).delete(categoryValue);
+      // wipe out the category if empty
+      if (this.filters.get(category).size === 0) {
+        this.filters.delete(category);
+      }
+    } else {
+      if (!this.filters.has(category)) {
+        this.filters.set(category, new Set<string>());
+      }
+      this.filters.get(category).add(categoryValue);
+    }
+  }
+  sortByAlphabet(orderedArray, orderMode): any {
+    orderedArray = orderedArray.sort(function (a, b) {
+      if (orderMode) {
+        return a.key > b.key ? 1 : -1;
+      } else  {
+        return a.key < b.key ? 1 : -1;
+      }
+    });
+    return orderedArray;
+  }
+
+  sortByCount(orderedArray, orderMode): any {
+    orderedArray = orderedArray.sort(function (a, b) {
+      if (a.value < b.value) {
+        return !orderMode ? 1 : -1;
+      } else if (a.value === b.value) {
+        return a.key > b.key ? 1 : -1;
+      } else {
+        return !orderMode ? -1 : 1;
+      }
+    });
+    return orderedArray;
+  }
+
+  sortCategoryValue (valueMap: any, sortMode: boolean, orderMode: boolean): any {
+    let orderedArray = <any>[];
+    valueMap.forEach(
```

Comment:
This looks like you're copying the array.  If it is, just use one of Object.assign or Object.create instead.